cpu/esp8266/periph/rtt.c

```diff
@@ -0,0 +1,201 @@
+        };
+        uint32_t val;
+    } intr;
+    uint32_t alarm;
+} frc2_struct_t;
+
+extern volatile frc2_struct_t frc2;
+
+typedef struct {
+    uint32_t alarm;         /**< alarm */
+    rtt_cb_t alarm_cb;      /**< alarm callback */
+    rtt_cb_t overflow_cb;   /**< overflow callback */
+    void *alarm_arg;        /**< argument for alarm callback */
+    void *overflow_arg;     /**< argument for overflow callback */
+} rtt_config_t;
+
+static rtt_config_t rtt_config = { };
+
+static uint32_t RTC_BSS_ATTR _rtt_counter_saved;
+static uint32_t RTC_BSS_ATTR _rtc_counter_saved;
+
+extern uint32_t pm_rtc_clock_cali_proc(void);
+extern uint32_t pm_rtc2usec(uint32_t rtc_cycles, uint32_t period);
+
+void IRAM rtt_cb(void *arg)
+{
+    if (rtt_config.alarm && rtt_config.alarm_cb) {
+        /* reset the alarm first */
+        rtt_cb_t alarm_cb = rtt_config.alarm_cb;
+        void * alarm_arg = rtt_config.alarm_arg;
+        rtt_clear_alarm();
+        frc2.alarm = 0;
+        /* call the alarm handler afterwards */
+        alarm_cb(alarm_arg);
+    }
+    else if (rtt_config.overflow_cb) {
+        rtt_config.overflow_cb(rtt_config.overflow_arg);
+    }
+}
+
+void rtt_init(void)
+{
+    DEBUG("%s saved rtt=%u rtc=%u\n",
+          __func__, _rtt_counter_saved, _rtc_counter_saved);
+
+    frc2.ctrl.clk_div = FRC2_CLK_DIV_256;
+    frc2.ctrl.reload = 0;
+    frc2.ctrl.intr_hold = 0;
+    frc2.ctrl.enable = 1;
+
+    if (_rtt_counter_saved || _rtc_counter_saved) {
+        /* if in init after power on, restore the RTT counter value */
+        uint32_t rtc_diff = RTC.COUNTER - _rtc_counter_saved;
+        uint32_t rtc_diff_us = pm_rtc2usec(rtc_diff, pm_rtc_clock_cali_proc());
+        uint32_t rtt_diff = RTT_US_TO_TICKS(rtc_diff_us);
+
+        frc2.load = _rtt_counter_saved + rtt_diff;
+
+        DEBUG("%s rtc_diff=%u rtt_diff=%u load=%u\n", __func__,
+              rtc_diff, rtt_diff, _rtt_counter_saved + rtt_diff);
+    }
+    else {
+        frc2.load = 0;
+        DEBUG("%s after power on\n", __func__);
+    }
+
+    /* emulate overflow interrupt */
+    frc2.alarm = 0;
+
+    ets_isr_attach (ETS_FRC2_INUM, rtt_cb, NULL);
+    ets_isr_unmask (BIT(ETS_FRC2_INUM));
+    DPORT.INT_ENABLE |= DPORT_INT_ENABLE_FRC2;
+}
+
+void rtt_poweron(void)
+{
+    /* power on simply reactivates the FRC2 counter */
+    frc2.ctrl.enable = 1;
+}
+
+void rtt_poweroff(void)
+{
+    /* power off simply deactivates the FRC2 counter */
+    frc2.ctrl.enable = 0;
+}
+
+void rtt_set_overflow_cb(rtt_cb_t cb, void *arg)
+{
+    /* there is no overflow interrupt, we emulate */
+    rtt_config.overflow_cb = cb;
+    rtt_config.overflow_arg = arg;
+}
+
+void rtt_clear_overflow_cb(void)
+{
+    /* there is no overflow interrupt, we emulate */
+    rtt_config.overflow_cb = NULL;
+    rtt_config.overflow_arg = NULL;
+}
+
+uint32_t rtt_get_counter(void)
+{
+    return frc2.count;
+}
+
+void rtt_set_counter(uint32_t counter)
+{
+    frc2.load = counter;
```

Comment:
How is `load` transferred into `count`?
Is this something the vendor library does?

Reply:
The `load` register is the usual way for ESP SoCs to set the counter value. You have register `count` that is RO to read the current value and a register `load` that is WO to set the value. It is the same as with GPIO set/clear registers.